### PR TITLE
feat(streaming): uniform row-id generate for table/stream source and make it global unique

### PR DESCRIPTION
**Pull request by yezizp2012**
## What's changed and what's your intention?

As title, generate row id in SourceExecutor for both table and stream source:
1. Init SourceExecutor with step info passed from meta when create and recovery.
2. Generate row id with embed epoch and different local stepped id in each source executor.

Storing step info in `AddMutation` seems weird here, make it run first and I think its time to refactor Mutation and Barrier definition now.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

**Comment by BugenZhao** [comment]
Please note that for chunks from the delete executor, we do not need to generate a row_id column. Maybe we need some mechanism to tell the source executor whether to generate an id.

**Comment by yezizp2012** [comment]
> Please note that for chunks from the delete executor, we do not need to generate a row_id column. Maybe we need some mechanism to tell the source executor whether to generate an id.

Yes, this is the reason why e2e failed. I'm working on it.

File: src/meta/src/barrier/command.rs [hunk 1]
```diff
@@ -64,8 +64,8 @@ pub enum Command {
 }
 
 impl Command {
-    pub fn checkpoint() -> Self {
-        Self::Plain(Mutation::Nothing(NothingMutation {}))
+    pub fn checkpoint(mutation: Option<Mutation>) -> Self {
+        Self::Plain(mutation.unwrap_or(Mutation::Nothing(NothingMutation {})))
```
Comment (MrCroxx) [inline src/meta/src/barrier/command.rs:68]: Remove `Option`, use `Mutation::Nothing` by default.

File: src/stream/src/executor/source.rs [hunk 2]
```diff
@@ -53,18 +55,56 @@ struct SourceReader {
...
+    actor_id: ActorId,
+
     column_ids: Vec<ColumnId>,
     schema: Schema,
+
     pk_indices: PkIndices,
 
-    /// current allocated row id
-    next_row_id: AtomicU64,
+    /// row id generator
```
Comment (MrCroxx) [inline src/stream/src/executor/source.rs:106]:
```suggestion
    /// Row id generator
```

**Review by MrCroxx — APPROVED** [review]
LGTM

File: src/meta/src/barrier/command.rs (hunk 1 above)
Comment (yezizp2012) [reply src/meta/src/barrier/command.rs:68]: Will fix with Mutation and Barrier refactor in next PR. 😂

File: src/stream/src/executor/source.rs [hunk 3]
```diff
@@ -216,7 +267,9 @@ impl SourceExecutor {
                 .position(|column_id| *column_id == row_id_column_id)
             {
                 let (ops, mut columns, bitmap) = chunk.into_inner();
-                columns[idx] = self.gen_row_column(columns[idx].array().len());
+                if ops.iter().any(|op| *op == Op::Insert) {
```
Comment (BugenZhao) [inline src/stream/src/executor/source.rs:270]: Is this okay? I'm not very clear how we handled delete records from the stream sources before...

File: src/meta/src/barrier/command.rs (hunk 1 above)
Comment (yezizp2012) [reply src/meta/src/barrier/command.rs]: Fixed with `plain_with` instead, keep `checkpoint` with nothing mutation.

File: src/stream/src/executor/source.rs (hunk 3 above)
Comment (BugenZhao) [reply src/stream/src/executor/source.rs:270]: How about letting the `Source` pass a flag of "whether/where to fill the row_id column" with each stream chunk?

Comment (yezizp2012) [reply src/stream/src/executor/source.rs:270]: This works for stream sources, deleting records from stream source only occurs when pk exist and there will be no row_id.

Comment (yezizp2012) [reply src/stream/src/executor/source.rs:270]: > How about letting the `Source` pass a flag of "whether/where to fill the row_id column" with each stream chunk?

It's a little bit weird too. 🤔

Comment (fuyufjh) [reply src/stream/src/executor/source.rs:270]: Can we just let the Source generate row-ids?

File: proto/data.proto [hunk 4]
```diff
@@ -148,8 +148,14 @@ message UpdateMutation {
   map<uint32, Actors> actors = 1;
 }
 
+message RowIdStepInfo {
```
Comment (TennyZhuang) [inline proto/data.proto:151]: How about `RowIdGenRule`?

File: src/stream/src/executor/source.rs (hunk 3 above)
Comment (yezizp2012) [reply src/stream/src/executor/source.rs:270]: Nope, for Delete message the row_id already queried and filled. Generate and refilling is not correct.

File: src/source/src/manager.rs [hunk 5]
```diff
@@ -42,7 +42,12 @@ const PROTOBUF_FILE_URL_SCHEME: &str = "file";
 #[async_trait]
 pub trait SourceManager: Debug + Sync + Send {
     async fn create_source(&self, table_id: &TableId, info: StreamSourceInfo) -> Result<()>;
-    fn create_table_source(&self, table_id: &TableId, columns: Vec<ColumnDesc>) -> Result<()>;
+    fn create_table_source(
+        &self,
+        table_id: &TableId,
+        columns: Vec<ColumnDesc>,
+        frontend_v2: bool,
```
Comment (TennyZhuang) [inline src/source/src/manager.rs:49]: Add a TODO so we can remove it after we remove support for frontend v1.

File: proto/data.proto (hunk 4 above)
Comment (yezizp2012) [reply proto/data.proto:151]: Cool, that sounds better. Will rename it.

**Comment by fuyufjh** [comment]
For append-only sources, the row-id can be generated either by SourceExecutor or Source itself, while for table source and CDC sources, it gets complex:

- Table source must provide the row-ids internally, considering updates & deletes
- CDC sources do not have row-ids. Instead, CDC tables must have PKs.

Therefore, I think it's more reasonable to let the source to decide how to generate all columns including row-id columns, instead of doing this in SourceExecutor.

File: src/stream/src/executor/source.rs (hunk 3 above)
Comment (TennyZhuang) [reply src/stream/src/executor/source.rs:270]: IMO there will be an append_only flag to SourceExecutor

**Comment by yezizp2012** [comment]
> For append-only sources, the row-id can be generated either by SourceExecutor or Source itself, while for table source and CDC sources, it gets complex:
> 
> * Table source must provide the row-ids internally, considering updates & deletes
> * CDC sources do not have row-ids. Instead, CDC tables must have PKs.
> 
> Therefore, I think it's more reasonable to let the source to decide how to generate all columns including row-id columns, instead of doing this in SourceExecutor.

For table source, the row-id is necessary:
- Insert, it should be generated and filled same with append-only sources.
- Others (Delete, Update), it will be queried and filled in the source side.
For CDS source, since it has Pk already and the row_id_idx will be None, no row-id will be generate and filled.

Besides, generate row-id in Source seems hard to guarantee unique in current implementation, since we use epoch and step info here to generate it. Unless we generate row-id in another way like timestamp(same with epoch) + worker_id + local_incremental_id. Do you think this solution acceptable?

File: src/stream/src/executor/source.rs (hunk 3 above)
Comment (yezizp2012) [reply src/stream/src/executor/source.rs:269]: append_only is not helpful for delete/update operation in table source.

**Comment by fuyufjh** [comment]
Just come up with an idea: 

how about letting the Source tells SourceExecutor which column is the row-id column? It should be an `Option<usize>`, where `None` means no row-id column needed to be assigned i.e. the PK is always provided by Source itself. For `Some(usize)`, for rows with `op == Op::Insert`, we fill the generated row-id in the specified row-id column.

**Comment by BugenZhao** [comment]
> Just come up with an idea:
> 
> how about letting the Source tells SourceExecutor which column is the row-id column? It should be an `Option<usize>`, where `None` means no row-id column needed to be assigned i.e. the PK is always provided by Source itself. For `Some(usize)`, for rows with `op == Op::Insert`, we fill the generated row-id in the specified row-id column.

+1. Currently, this info is fixed for a specific source.

**Pull request closed without merging**